### ISSUE TITLE
Eppy suffers from performance issues for larger IDF models

### ISSUE THREAD

**Issue by samuelduchesne** [santoshphilip/eppy#303]
I have found out that loading some of the heftier models can take a while to load.

I am investigating possible fixes

**Comment by santoshphilip** [comment]
ha ! I was not aware of that.
I know that the IDD takes time to load. Did not know that about the IDF.
Are you running a profiler on it, to see where it is taking time?

Let us know what you discover

**Comment by samuelduchesne** [comment]
Yes, I'm running yappi to figure out which function is taking a lot of time. Then I use line_profile to profile line by line.

I found out that "only_legalcharrs" is very heavy, especially since it can be called millions of times.

https://github.com/santoshphilip/eppy/blob/98e58583dce6c0fcec9c7b1ff1142bae0a67ddc7/eppy/bunchhelpers.py#L17-L20

# Possible solution

A simple fix is to encode and decode the string:

```
def faster_onlylegalchar(name):
    return name.encode('ascii', 'ignore').decode()
```

It is much faster (~7x faster). This translated in a 30% speed increase when parsing a large IDF file.

```
>>> %timeit onlylegalchar("This is a non-ascii: Æ")
2 µs ± 13.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

>>> %timeit faster_onlylegalchar("This is a non-ascii: Æ")
272 ns ± 1.29 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

The result does keep characters the `onlylegalchars` would keep such as dashes `-` and `:`

```
>>> onlylegalchar("This is a non-ascii: Æ")
'This is a nonascii '

>>> faster_onlylegalchar("This is a non-ascii: Æ")
'This is a non-ascii: '
```

Let's see if the CI is happy...

**Comment by santoshphilip** [comment]
Apparently CI is unhappy
https://ci.appveyor.com/project/santoshphilip/eppy/builds/34998766

I am making some changes in response to #287 that will fix some of that unhappiness. I'll merge those changes in a couple of days.

**Comment by samuelduchesne** [comment]
Another lag is created by these two decorators: `make_idd_index` and `extractidddata` which essentially translates into 3 calls to `extractidddata`. Rewriting this function would speed up significantly the `idfreader1` function.

https://github.com/santoshphilip/eppy/blob/98e58583dce6c0fcec9c7b1ff1142bae0a67ddc7/eppy/EPlusInterfaceFunctions/parse_idd.py#L143-L379

**Comment by santoshphilip** [comment]
I suspected `make_idd_index` was a culprit.

I used `make_idd_index` as a way of speeding up searches thru the file. 
- like finding walls in a zone
- the idea behind it was to have something like indexing in databases (mongodb was the inspiration)
- I don't think it speeded things up much.

removing it is an option

regarding `embedgroupdata`, I have look into it to see what `embedgroupdata` does and why it is useful. I don't recall off the top of my head

**Comment by santoshphilip** [comment]
Do you have a sample large file I can play with.

I also want to test it with [eppy3000](https://github.com/pyenergyplus/eppy3000). So far `eppy3000` has turned out to be 5-10 faster than `eppy`

**Comment by santoshphilip** [comment]
I updated branch `develop` with some changes. 
It may fix some of your CI issues.

**Comment by samuelduchesne** [comment]
> Do you have a sample large file I can play with.
> 
> I also want to test it with [eppy3000](https://github.com/pyenergyplus/eppy3000). So far `eppy3000` has turned out to be 5-10 faster than `eppy`

Yes I need to check that out! Happy to see you are moving to epjson

**Comment by samuelduchesne** [comment]
> I updated branch `develop` with some changes.
> It may fix some of your CI issues.

latest changes on #304 seem to pass all tests.

I implemented a regex instead of encode/decode. It is not as fast, but at least follows the same behavior as the original `onlylegalchars` function.

**Comment by samuelduchesne** [comment]
#304 also merges the 2 decorators inside the `extractidddata`, which prevents calling it 3 times.
